### DISCUSSION TITLE
Feature Request: Generate fields for "uses" Mappers used in @BeforeMapping/@AfterMapping methods

### DISCUSSION THREAD

**Discussion by mcmasjc3** [mapstruct/mapstruct#3776] [Ideas]
I have a Mapper which calls @AfterMapping methods that invoke other Mappers. The manual says, "The generated mapper will inject classes defined in the uses attribute if MapStruct has detected that it needs to use an instance of it for a mapping."

 Apparently, this is only true for the main mapping method itself. It does not detect Mappers used by @AfterMapping methods, even if defined in an abstract class where they could use them in fields if MapStruct created them.

 This means I have to pass any Mappers used by those methods as @Context parameters. This makes calling the Mapper more cumbersome. It would be really nice if it could generate fields for the Mappers used by these other methods.

**Comment by filiphr** [comment]
@mcmasjc3 why can't you just use an abstract class with fields that you want to use in your `@BeforeMapping` / `@AfterMapping`?

**Reply by mcmasjc3** [reply]
I tried that. I am using Spring constructor injection to instantiate my Mapper. MapStruct does not create constructor parameters to populate the fields, because it does not believe they are used.

**Reply by filiphr** [reply]
You need to use either a field `@Autowired` or a setter `@Autowired` to inject what you need. At least until we do PR #3546

**Reply by mcmasjc3** [reply]
Thank you. For the moment I created a wrapper class to contain the Mappers I need to keep the parameter count down. @Autowired fields make test ing a lot harder. I could use @Autowired setters, but I really dislike them. I will stick with the wrapper class for now.

**Comment by RichMacDonald** [comment]
Even though I am a spring user, I find the mapstruct `public static final MyMapper INSTANCE = Mappers.getMapper(MyClass.class)` works so well that it is the way to go. Startup configuration is hard enough, and this is one less thing to write explicit code for.

1.  I use spring configuration classes to instantiate the context instances but never the mappers.
1. Never `@Autowired`. Autowired is evil for testing. It should not be used anywhere.
2. Every Mapping class (I use interfaces) has its public INSTANCE field. So any Before/AfterMapping code that delegates to another Mapper just refers to that INSTANCE.
3. If the mappers need some extra startup configuration (e.g., I register all Mapping instances in a HashMap keyed by class in order to manage class hierarchy lookup), this can be handled by code inside static blocks.
4. Because there is only one instance for unit testing, more things get pushed to the context. But this is a good thing. The Mapper code only handles "strict" mapping and asks the context for everything else that it needs. This makes unit testing straightforward: Test the Mapper with mocked contexts. Test the context with other mocks that have nothing to do with the Mapper..

**Comment by mcmasjc3** [comment]
I avoid @Autowired fields and setters like the plague. I use constructor
injection in order to make unit testing easier. That is why I want
MapStruct to generate the fields (and constructor parameters) for the
Mappers called by @BeforeMapping/AfterMapping methods. If I add the fields
to the abstract class myself, then they are not injectable through the
constructor.

Jim McMaster
***@***.***

**Reply by RichMacDonald** [reply]
I understand, and perhaps this is an oversight in the framework. But honestly I don't like your approach. I want my mappers to know only how to map _provided_ Objects and primitives from one type to another recursively. That is what they do well. Anything else is "leaky responsibility" and should belong to the Context.

Knowing where to get extra information, side-effects, etc, are all necessary parts of a real-world model mapping framework. The mapper classes are just one "dumb as possible" part. In practice, a "not so dumb" part that is hard enough without including all the "external issues".

Forcing the Mapper to have no instance attributes and only the one static attribute INSTANCE makes it _really_ test-friendly. Whatever can't fit into that requirement goes into the Context.

And you'll make the Context an interface as well, so it is easy to mock. That makes unit-testing even more flexible.

Can even use different Context interfaces for each mapping method, so you keep your Context instances lean. Or have one Context class implement N Context interfaces if that works better.

We are discussing how to split the inherent complexity of the mapping problem between the Mappers and the Contexts. The Mappers have a few rules and restrictions which the Contexts don't. Yes, you pay for it by an extra parameter in all your mapping methods, but what you gain is a Context that represents your "interface to the world". Annoying boilerplate yes, but it remains constant as complexity increases. And that is the winning aspect, which is always "how can I continue to add, test and manage small things without increasingly severe coupling".

I find I can write straightforward unit tests for each field mapping, one at a time, using mostly mocked test classes. And then one end-to-end test to prove all the individual mappings work collectively.

I might be overthinking my response to you. All experienced opinion matters and we're debating high-level preferences and theory without concrete examples. Happy to continue this, if you want to work on some specific examples.

Rich MacDonald
